docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+
+The gas service handles tracking of gas consumptions at the block and transaction level and also allows
+for overriding the gas meter passed to child calls:
+```go
+package gas
+
+type Service interface {
+	GetMeter(context.Context) Meter
+	GetBlockMeter(context.Context) Meter
+	WithMeter(ctx context.Context, meter Meter) context.Context
+	WithBlockMeter(ctx context.Context, meter Meter) context.Context
+}
+```
+
+#### Inter-module Client
+
+Runtime module implementations should provide an instance of the `appmodule.RootInterModuleClient` interface which
+will allow modules to send messages to and make queries against other modules as described in [ADR 033](./adr-033-protobuf-inter-module-comm.md).
+
+```go
+type InterModuleClient interface {
+    grpc.ClientConnInterface
+	
+	// Address is the ADR-028 address of this client against which messages will be authenticated.
+	Address() []byte
+}
+
+type RootInterModuleClient interface {
+	InterModuleClient
+	
+    DerivedClient(key []byte) InterModuleClient
+}
+```
+
+The `RootInterModuleClient` allows a module to make inter-module calls using its root [ADR 028](./adr-028-public-key-addresses.md)
+address (as returned by the `Address` method).
+
+This client can be used with both `MsgClient`'s and `QueryClient`'s and the protobuf option
+`cosmos.msg.v1.service` will allow the router to detect whether a given service is a query or msg service and
+route requests appropriately.
+
+Module clients that use [ADR 028](./adr-028-public-key-addresses.md) derived module addresses
+can be created using the `DerivedClient` method.
+
+By default, all `Msg`'s registered using the core API will be available for inter-module routing. Queries will only
+be exposed for inter-module routing if their `rpc` methods have the protobuf option `cosmos.query.v1.internal`
+set to true.
+
+The inter-module router will provide a unified interface for sending `Msg`'s and queries
+which ensures all required pre-processing steps are uniformly executed by all modules which need such functionality.
+
+Runtime modules may choose to allow different types of "admin" access to individual modules which may allow for
+bypassing certain authentication checks. This could allow `x/authz` to allow sending messages on behalf of all
+addresses, for example.
+
+#### `appmodule.Service` Bundle Service
+
+To allow modules to declare a dependency on a single "bundle" service, runtime modules should provide an implementation
+of the `appmodule.Service` interface:
+
+```go
+package appmodule
+
+type Service interface {
+	store.KVStoreService
+	store.MemoryStoreService
+	store.TransientStoreService
+	event.Service
+	blockinfo.Service
+	gas.Service
+    RootInterModuleClient
+}
+```
+To maintain API compatibility, if new core services are added, a new `cosmossdk.io/coreappmodule/v2.Service` should
+be added which extends this service and bundles the new core services, ex:
```

Comment:
Historically how often has this happened? Seems like a big downside of this approach.

Reply:
Not really sure, but why is it a big downside? We just bump the version

Reply:
As a user of the Cosmos SDK it would make upgrading to a new core API version more of a refactor and therefore more cumbersome, no?

Reply:
Why? They just need to change the import

Reply:
Changing a `go.mod` import is less work than changing a `go.mod` import and import statements across `.go.` files, which is required since the package is now `v2`.  Maybe "big" drawback should be modified to "minor" drawback.

Reply:
Why do you need to change so many imports? You only need to change imports that require the new service

Reply:
On reflection I guess it's no more work than updating a method signature to support a new service.  In fact it's probably easier to change an import statement than a func sig.  Am I thinking about this right?

Reply:
I think so